Fix VectorTools::interpolate for parallel computations

gassmoeller opened the pull request:
This hopefully fixes the rest of the issues in geodynamics/aspect#1936 that are not addressed in #5313. I am not quite sure the fix is correct for all cases though (I am not familiar with exotic elements / hp / ...), so I would like to ask for a careful review. I can add a test when I am sure this is correct.

The error message that is fixed by this PR is:
```
1: --------------------------------------------------------
1: An error occurred in line <603> of file </home/fraters/deal.ii/deal.ii_2017-09-26/dealii/source/lac/trilinos_vector.cc> in function
1:     dealii::TrilinosScalar dealii::TrilinosWrappers::MPI::Vector::operator()(dealii::TrilinosWrappers::MPI::Vector::size_type) const
1: The violated condition was: 
1:     false
1: Additional information: 
1:     You tried to access element 30 of a distributed vector, but this element is not stored on the current processor. Note: There are 72 elements stored on the current processor from within the range 90 through 161 but Trilinos vectors need not store contiguous ranges on each processor, and not every element in this range may in fact be stored locally.
1: 
1: Stacktrace:
1: -----------
1: #0  /home/fraters/deal.ii/deal.ii_2017-09-26/dealii/build/lib/libdeal_II.g.so.9.0.0-pre: dealii::TrilinosWrappers::MPI::Vector::operator()(unsigned int) const
1: #1  /home/fraters/deal.ii/deal.ii_2017-09-26/dealii/build/lib/libdeal_II.g.so.9.0.0-pre: void dealii::VectorTools::interpolate<2, 2, dealii::TrilinosWrappers::MPI::BlockVector, dealii::DoFHandler>(dealii::Mapping<2, 2> const&, dealii::DoFHandler<2, 2> const&, dealii::Function<2, dealii::TrilinosWrappers::MPI::BlockVector::value_type> const&, dealii::TrilinosWrappers::MPI::BlockVector&, dealii::ComponentMask const&)
1: #2  /home/fraters/aspect/aspect_2017-09-26/aspect/build/aspect: aspect::Simulator<2>::set_initial_temperature_and_compositional_fields()
1: #3  /home/fraters/aspect/aspect_2017-09-26/aspect/build/aspect: aspect::Simulator<2>::run()
1: #4  /home/fraters/aspect/aspect_2017-09-26/aspect/build/aspect: main
1: --------------------------------------------------------
1: 
1: Calling MPI_Abort now.
1: To break execution in a GDB session, execute 'break MPI_Abort' before running. You can also put the following into your ~/.gdbinit:
1:   set breakpoint pending on
1:   break MPI_Abort
1:   set breakpoint pending auto
1: make[3]: *** [output-quick_mpi/screen-output] Error 255
1: make[2]: *** [CMakeFiles/quick_mpi.screen-output.diff.dir/all] Error 2
1: make[1]: *** [CMakeFiles/tests.quick_mpi.dir/rule] Error 2
1: make: *** [tests.quick_mpi] Error 2
1: *** "Test quick_mpi failed": ***
1: [ 22%] Built target quick_mpi
1: [ 33%] Generating output-quick_mpi/screen-output
1: CMakeFiles/quick_mpi.screen-output.diff.dir/build.make:76: recipe for target 'output-quick_mpi/screen-output' failed
1: CMakeFiles/Makefile2:164: recipe for target 'CMakeFiles/quick_mpi.screen-output.diff.dir/all' failed
1: CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/tests.quick_mpi.dir/rule' failed
1: Makefile:142: recipe for target 'tests.quick_mpi' failed
1: 
1: CMake Error at /home/fraters/aspect/aspect_2017-09-26/aspect/tests/run_test.cmake:14 (MESSAGE):
1:   *** Test aborted.
1: 
1: 
1/1 Test #1: quick_mpi ........................***Failed    0.43 sec
```

The problem here is that the interpolation in `VectorTools::interpolate` is only done for locally owned cells, but for all dofs on these cells. In some cases locally owned cells are associated with non locally owned dofs (e.g. at the boundary to ghost cells). This function currently tries to write into these dofs from both processes, which triggers an assertion of the parallel vector. Skipping these dofs should fix this, and should do no harm in many cases (they are interpolated on the owning process and then transferred during the call to `compress`). However, if the interpolated function is different on both cells for the same point, than this will lead to wrong results (there is no averaging of the values). I am not sure how to work around this though. @tamiko or somebody else, do you have an idea?

tjhei commented:
I don't think this is the correct solution: We still want to average all values from all neighboring cells. With your fix we only average values for own dofs from locally owned cells.

tjhei commented:
I think you can fix this by moving your check if the vector entry is locally owned to the last ``else`` block in the loop, so only do "// If a component is ignored, simply copy all dof values from the vector "vec"" if we are the owner of that DoF.

gassmoeller commented:
Good point, I was not aware that we can not read from non locally owned elements, but write into them (I suppose into a buffer that is communicated during `compress`?). These low-level things are still somewhat new to me. 
Then this new version should fix all issues, averaging is still performed, and only dofs that are not locally owned and are not selected for interpolation are ignored, which is fine.
I will try to come up with a test tomorrow, but unless there are other issues this should work.

masterleinad commented:
/run-tests

bangerth commented:
Then I don't understand why you got that error in the first place -- I thought that happened because you were writing into vector elements you don't own? @tjhei is of course correct that that is allowed (why did I not think of this when we were standing at the board). Where exactly did the error happen then?

gassmoeller commented:
You are right, I did not check that carefully enough. It is the read that fails, because some dof of the ghost cell is not available on the local process (it is a fully distributed vector). Does that make more sense?

bangerth commented:
Yes, but can you show the place where the read fails? I thought the function carefully only ever writes into the vector.

gassmoeller commented:
It is hard to proof, because my debugger always exits instead of giving me a backtrace, but it is in the `operator () const` function of the trilinos vector, and the only reasonable place I can see is in the `::dealii::internal::ElementAccess<VectorType>::get` function right below the newly inserted lines. Apart from that there are only `::dealii::internal::ElementAccess<VectorType>::add` functions inside this if-else block, and it definitely fails inside that block.

bangerth commented:
If you read the error message before the program aborts, then there is a hint at what you need to tell gdb to actually stop at the place where something happens. It boils down to setting a breakpoint on `MPI_Abort` :-) This should get you a backtrace, and a way to debug things even in parallel.

tjhei commented:
``::get()`` reads all values on owned cells and will trigger an error for non-owned DoFs. Writes to entries that are not ours is of course allowed. Just skipping the else block is correct, because only one processor needs to do this computation (and will do it now). You can not skip the whole loop, because then averaging between processor boundaries is incomplete.

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -448,6 +449,10 @@ namespace VectorTools
                 {
+                  // Skip the operation below if we are not dealing with a local dof
+                  if (!locally_owned_dofs.is_element(dofs_on_cell[i]))
+                    continue;
```
I think this would be easier to read if you turned this around and used
```
  if (is locally owned)
    ...do operation...
```

tjhei commented:
> So, @tjhei, are you saying that this patch is correct then?

I liked the version from 17 hours ago (after my first comment). I am not sure about the current one just pushed.

gassmoeller commented:
The reason I changed it was that in the earlier version both processes (the owning and ghost neighbor) added 1 to the weights, but the ghost owner did not add a value to the interpolation vector, this sounded wrong to me. I only moved the `add` to the weights to the sample place where the `add` to the interpolation vector is happening. I also added a test that fails on master, but works on this branch.

bangerth commented:
OK to merge once the tester is happy.

bangerth commented:
Thanks a lot for figuring this out!

gassmoeller commented:
Is there a simple way to see from the vector norm if the test result is correct? If not, I would like to modify the test to actually integrate over the interpolated function to see that the result is correct.

bangerth commented:
We usually interpolate a function onto the FE space that's actually in this space -- i.e., the interpolation is the identity operation. You can then use `VectorTools::integrate_difference` and check that the error is zero.

gassmoeller commented:
Thanks for the suggestions, I update the test to produce meaningful output, and checked that it produces the expected result. This is now ready from my side.
